spec/syntax.md

```diff
@@ -526,18 +528,73 @@ unrecognized reserved sequences have no meaning and MAY result in errors during
 
 ```abnf
 reserved       = reserved-start reserved-body
-reserved-start = "!" / "@" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"
+reserved-start = "!" / "#" / "%" / "*" / "<" / ">" / "/" / "?" / "~"
 
 reserved-body  = *( [s] 1*(reserved-char / reserved-escape / quoted))
 reserved-char  = %x00-08        ; omit HTAB and LF
                / %x0B-0C        ; omit CR
                / %x0E-19        ; omit SP
-               / %x21-5B        ; omit \
+               / %x21-3F        ; omit @
+               / %x41-5B        ; omit \
                / %x5D-7A        ; omit { | }
                / %x7E-D7FF      ; omit surrogates
                / %xE000-10FFFF
 ```
 
+### Expression Attributes
+
+An **_<dfn>expression attribute</dfn>_** is a key-value pair
+containing a named argument that is attached to the _expression_ as a whole.
+They are intended to convey information about the _expression_
+that is not necessarily applicable as a _function_ _option_.
+For example, the `@locale` attribute overrides the locale of the _message_ for its _expression_.
+Attributes like `@can-delete` and `@translate` might not have any impact on formatting,
+acting instead as instructions to translators.
+
+_Expression attributes_ are prefixed by a U+0040 COMMERCIAL AT `@` sign,
+but otherwise use the same syntax as _options_.
+
+The _name_ of an _expression attribute_ MUST either be one of the following,
+or it MUST contain at least one U+002D HYPHEN-MINUS `-` character:
+
+- `language`: SHOULD resolve to a sequence of IETF BCP 47 language tags;
+  see _expression attribute resolution_.
+- `translate`: SHOULD have a literal value of either `yes` or `no`.
+
+To ensure future extensibility of this specification,
+all other _expression attributes_ with a _name_ not containing `-` MUST be ignored.
```

Comment:
How about just allowing unrecognized attributes to be ignored?

Reply:
Because then we could not later assign a meaning to them without potentially breaking user code.

Reply:
Reserving a namespace is a separate thing from ignoring unrecognized values. These can work in tandem. An MF2.0 implementation would ignore the unrecognized value `@foo` that MF2.1 unreserves, as well as `@_bar` that some other proprietary implementation is using.

Alternatively (and as I suggested elsewhere), as long as the values are syntactically correct, the implementation can just parse the values without imputing meaning to them and pass them. It only has to apply validation rules to attributes it "knows" (and do we define an error for that???)

Reply:
Consider the following hypothetical, in a world where we allow implementations to assign their own meaning to any attributes not explicitly defined in the spec like `@locale` and `@translate`:

Let's say that an implementation defines `@fallback` to provide a value that should be used for a placeholder, should its formatting fail for whatever reason. This doesn't really make sense as a function option, because it's covering the behaviour of what happens if the function throws an error. Other implementations could just ignore this attribute, and use the spec-defined fallback representation.

Then let's say that in a 2.1 version of the MF spec we'd like to define `@fallback` to mean a fallback locale to use, if the function doesn't support the message locale, or the one set by `@locale`. This hypothetical attribute would have the same semantics as `@locale`, but it would append values to the language priority list rather than replacing it.

But in this world we could not do that without potentially breaking userspace, because an implementation as defined above could exist, along with a pre-existing body of messages using `@fallback` with a completely different meaning. In this case, the meanings of "fallback" are quite different, but even should they align, it's very likely that the detailed implementation and requirements could differ for the attributes. In other words, we would not be able to define _any_ future attributes in the spec without a danger of breaking userspace.